main/main.c

```diff
@@ -15,42 +13,328 @@
+static void handler_unknown(void)
+{}
+
+static void packet_create_header(packet_t *packet, DataType type)
+{
+    packet->version = (CONFIG_PACKET_VERSION << 4); // Version and 4-bit reserved
+    packet->id = (CONFIG_CLASS_ID << 4) | CONFIG_DEVICE_ID; // This should be unique
+    packet->dataType = type;
+
+    /*
+     * TEMPORARY: update msg counter and ID for debug purpose.
+     */
+    packet->msgCount = node.radio.msg_counter++;
+    packet->msgID = node.radio.msg_id;
+    if (!node.radio.msg_counter)
+    {
+    	node.radio.msg_id++;
+    }
+}
+
+static void sender_unknown(void)
+{
+	packet_create_header(&packet, SMS);
+	sprintf((char * restrict)packet.data, "Unknown:0x%x", node.status);
+	lora_send(&packet);
+}
+
+static void sender_bme280(void)
+{
+    if (ESP_FAIL == bme280_read_temperature(&node.sensor.bme280.temp_raw))
+    {
+    	node.sensor.bme280.comms_err_cnt++;
+    }
+    if (ESP_FAIL == bme280_read_pressure(&node.sensor.bme280.press_raw))
     {
-        ESP_LOGE(TAG_MAIN, "BME280 initialization failed");
-        // not sure if this should restart, but also not sure what else to do here?
-        esp_restart();
+    	node.sensor.bme280.comms_err_cnt++;
     }
+    if (ESP_FAIL == bme280_read_humidity(&node.sensor.bme280.hum_raw))
+    {
+    	node.sensor.bme280.comms_err_cnt++;
+    }
+
+    packet_create_header(&packet, BME280);
 
-    bme_rc += bme280_set_oversamp(BME280_OVERSAMP_16X, BME280_OVERSAMP_16X, BME280_OVERSAMP_16X);
-    bme_rc += bme280_set_settings(STANDBY_10MS, BME280_FILTER_COEFF_16, BME280_NORMAL_MODE);
+    // Convert raw sensor readings
+	packet.data[0] = (int8_t)(node.sensor.bme280.temp_raw * 2);             // Scale temperature for higher precision and fit into int8_t
+	packet.data[1] = (int8_t)((node.sensor.bme280.press_raw / 100) - 1000); // Convert Pa to hPa, subtract 1000
+	packet.data[2] = (uint8_t)node.sensor.bme280.hum_raw;                   // Fit humidity into uint8_t
 
-    if (ESP_OK != bme_rc)
+	// Log the packet data before sending
+	ESP_LOGI(pcTaskGetName(NULL), "Temperature: %.2f C (scaled to %d)", node.sensor.bme280.temp_raw, packet.data[0]);
+	ESP_LOGI(pcTaskGetName(NULL), "Pressure: %.2f hPa (stored as %d)", node.sensor.bme280.press_raw / 100, packet.data[1]);
+	ESP_LOGI(pcTaskGetName(NULL), "Humidity: %.2f %% (stored as %u)", node.sensor.bme280.hum_raw, packet.data[2]);
+
+    lora_status_t send_status = lora_send(&packet);
+
+    if (LORA_OK != send_status)
     {
-        ESP_LOGE("BME", "BME280 settings failed");
-        // not sure if this should restart, but also not sure what else to do here?
-        esp_restart();
+        ESP_LOGE(TAG, "Packet send failed");
     }
+    else
+    {
+        ESP_LOGI(TAG, "Packet sent successfully");
+    }
+
 }
-#endif
 
-#if CONFIG_LORA_RECEIVER
-static void task_rx(void *pvParameters);
-#define MSG_BUFFER_SIZE 128
-#endif
+static void sender_status(void)
+{
+    packet_create_header(&packet, STATUS);
+    packet.data[0] = node.status;

```

Comment:
Maybe check the data array size before? Are you sure that it has at least 5 bytes always? What if someone will change the size of the data?

Reply:
Data by definition data[] is of size 59 (for SMS) so 5 is guaranteed. Need to think about how to make all those checks.